### PR TITLE
[aot] [refactor] Refactor AOT field API for Vulkan

### PR DESCRIPTION
**Pull request by qiao-bo**
Related issue = #3642

Add field API `auto x_field = vk_module->get_field("place");`

### CODE REVIEW

File: taichi/backends/vulkan/aot_module_loader_impl.cpp [hunk 1]
```diff
@@ -12,6 +12,31 @@ namespace {
...
 
+class FieldImpl : public aot::Field {
+ public:
+  explicit FieldImpl(VkRuntime *runtime, aot::CompiledFieldData field)
+      : runtime_(runtime), field_(field) {
+  }
+
+  void copy_to_host_buffer(uint64_t *dst_host) override {
+    // since kernel.num_snode_trees = 1, we always locate the first root
+    int root_id = 0;
```
Comment (k-ye) [inline taichi/backends/vulkan/aot_module_loader_impl.cpp:23]: nit: prefer `constexpr` whenever it is a compile-time constant

File: taichi/backends/vulkan/aot_module_loader_impl.cpp [hunk 2]
```diff
@@ -12,6 +12,31 @@ namespace {
 
 using KernelHandle = VkRuntime::KernelHandle;
 
+class FieldImpl : public aot::Field {
+ public:
+  explicit FieldImpl(VkRuntime *runtime, aot::CompiledFieldData field)
```
Comment (k-ye) [inline taichi/backends/vulkan/aot_module_loader_impl.cpp:17]:
```suggestion
  explicit FieldImpl(VkRuntime *runtime, const aot::CompiledFieldData &field)
```

**Review by k-ye — COMMENTED** [review]
Just a few nits, thanks!

File: taichi/backends/vulkan/aot_module_loader_impl.cpp [hunk 3]
```diff
@@ -69,6 +94,26 @@ class AotModuleImpl : public aot::Module {
   }
 
  private:
+  std::unique_ptr<aot::Field> make_new_field(const std::string &name) override {
+    aot::CompiledFieldData field;
```
Comment (ailzhang) [inline taichi/backends/vulkan/aot_module_loader_impl.cpp:98]: shall we put this directly in the `get_field` function?

**Review by ailzhang — COMMENTED** [review]
Thanks!

File: taichi/backends/vulkan/runtime.cpp [hunk 4]
```diff
@@ -591,15 +591,32 @@ void VkRuntime::add_root_buffer(size_t root_buffer_size) {
     root_buffer_size = 4;  // there might be empty roots
   }
   std::unique_ptr<DeviceAllocationGuard> new_buffer =
-      device_->allocate_memory_unique(
-          {root_buffer_size,
-           /*host_write=*/false, /*host_read=*/false,
-           /*export_sharing=*/false, AllocUsage::Storage});
+      device_->allocate_memory_unique({root_buffer_size,
+                                       /*host_write=*/true, /*host_read=*/true,
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply taichi/backends/vulkan/runtime.cpp:595]: is there a better way? without this change root_buffer cannot be mapped for discrete GPUs.

File: taichi/backends/vulkan/aot_module_loader_impl.cpp (hunk 3 above)
Comment (qiao-bo) [reply taichi/backends/vulkan/aot_module_loader_impl.cpp:98]: sure

File: tests/cpp/aot/aot_save_load_test.cpp [hunk 5]
```diff
@@ -146,8 +146,11 @@ TEST(AotSaveLoad, Vulkan) {
...
 
-  // auto x_field = vk_module.get_field("x");
-  // EXPECT_TRUE(x_field);
-  // x_field.copy_to(/*dst=*/x.get());
+  // Retrieve data
+  auto x_field = vk_module->get_field("place");
+  EXPECT_TRUE(x_field);
+
+  std::vector<uint64_t> x(root_size);
+  x_field->copy_to_host_buffer(&x[0]);
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply tests/cpp/aot/aot_save_load_test.cpp:154]: You mean to check the correctness of the returned data?

**Comment by qiao-bo** [comment]
Update: removed the `copy_to_host_buffer` API to avoid perf regression.

File: tests/cpp/aot/aot_save_load_test.cpp [hunk 6]
```diff
@@ -146,8 +146,8 @@ TEST(AotSaveLoad, Vulkan) {
   ret_kernel->launch(&host_ctx);
   vulkan_runtime->synchronize();
 
-  // auto x_field = vk_module.get_field("x");
-  // EXPECT_TRUE(x_field);
-  // x_field.copy_to(/*dst=*/x.get());
+  // Retrieve data
+  auto x_field = vk_module->get_field("place");
+  EXPECT_TRUE(x_field);
```
Comment (k-ye) [inline tests/cpp/aot/aot_save_load_test.cpp:151]:
```suggestion
  EXPECT_NE(x_field, nullptr);
```

https://google.github.io/googletest/reference/assertions.html

File: taichi/backends/vulkan/aot_module_loader_impl.cpp [hunk 7]
```diff
@@ -12,6 +12,17 @@ namespace {
 
 using KernelHandle = VkRuntime::KernelHandle;
 
+class FieldImpl : public aot::Field {
```
Comment (k-ye) [inline taichi/backends/vulkan/aot_module_loader_impl.cpp:15]: nit: later we can add basic interfaces like field size, etc.

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**